src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/OleVariant.cs

```diff
@@ -0,0 +1,560 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices.Marshalling
+{
+    /// <summary>
+    /// A type that represents an OLE VARIANT in managed code.
+    /// </summary>
+    [StructLayout(LayoutKind.Explicit)]
+    public struct OleVariant : IDisposable
+    {
+        private const VarEnum VT_VERSIONED_STREAM = (VarEnum)73;
+#if DEBUG
+        static unsafe OleVariant()
+        {
+            // Variant size is the size of 4 pointers (16 bytes) on a 32-bit processor,
+            // and 3 pointers (24 bytes) on a 64-bit processor.
+            int variantSize = sizeof(OleVariant);
+            if (IntPtr.Size == 4)
+            {
+                Debug.Assert(variantSize == (4 * IntPtr.Size));
+            }
+            else
+            {
+                Debug.Assert(IntPtr.Size == 8);
+                Debug.Assert(variantSize == (3 * IntPtr.Size));
+            }
+        }
+#endif
+
+        // Most of the data types in the Variant are carried in _typeUnion
+        [FieldOffset(0)] private TypeUnion _typeUnion;
+
+        // Decimal is the largest data type and it needs to use the space that is normally unused in TypeUnion._wReserved1, etc.
+        // Hence, it is declared to completely overlap with TypeUnion. A Decimal does not use the first two bytes, and so
+        // TypeUnion._vt can still be used to encode the type.
+        [FieldOffset(0)] private decimal _decimal;
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct TypeUnion
+        {
+            public ushort _vt;
+            public ushort _wReserved1;
+            public ushort _wReserved2;
+            public ushort _wReserved3;
+
+            public UnionTypes _unionTypes;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct Record
+        {
+            public IntPtr _record;
+            public IntPtr _recordInfo;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        private struct Blob
+        {
+            public int _size;
+            public IntPtr _data;
+        }
```

Comment:
Is there a reason we can't use `void*` here? I think we prefer that when expressing these things.

Reply:
We'd end up having to cast it to `IntPtr` anyway in its only usage, so it doesn't feel worthwhile to switch.